src/rules/no-incompatible-branch-types.coffee

```diff
@@ -0,0 +1,63 @@
+ts = require 'typescript'
+
+# getSourceFileOfNode = (tsNode) ->
+#   while node and node.kind isnt ts.SyntaxKind.SourceFile
+#     node = node.parent
+#   node
+
+module.exports =
+  meta:
+    docs:
+      description:
+        'Flag incompatible early-return types in branch()/branchPure()'
+      category: 'Possible Errors'
+      recommended: yes
+    schema: []
+
+  create: (context) ->
+    {parserServices} = context
+    return {} unless parserServices?.hasFullTypeInformation
+
+    typeChecker = parserServices.program.getTypeChecker()
+
+    CallExpression: (node) ->
+      {callee, parent} = node
+      return unless (
+        callee?.type is 'Identifier' and
+        callee.name in [### 'renderNothing', ### 'returns']
+      )
+
+      return unless (
+        parent.type is 'CallExpression' and
+        parent.callee.type is 'Identifier' and
+        parent.callee.name in ['branch' ### , 'branchPure'###]
+      )
+
+      {arguments: [returnsCallbackNode]} = node
+      return unless returnsCallbackNode?
+
+      tsReturnsCallbackNode = parserServices.esTreeNodeToTSNodeMap.get(
+        returnsCallbackNode
+      )
+      tsReturnsCallbackType = typeChecker.getTypeAtLocation(
+        tsReturnsCallbackNode
+      )
+      [signature] = typeChecker.getSignaturesOfType(
+        tsReturnsCallbackType
+        ts.SignatureKind.Call
+      )
+      return unless signature?
+      returnType = typeChecker.getReturnTypeOfSignature signature
+      returnTypeString = typeChecker.typeToString returnType
+      return if returnTypeString in ['null', 'Element', 'Element | null']
```

Comment:
Had to dig a bit into what a JSX element return type looks like as far as Typescript type checking (and how to test that)

Per [the docs](https://www.typescriptlang.org/docs/handbook/jsx.html#the-jsx-result-type), I was seeing `any` as the return type of a function return a JSX element (in testing) until I added a declaration for `JSX.Element` to the tests, at which point it reported `Element` as the return type (which is also what I saw when running this in the context of a "real" React Native project)